### PR TITLE
fix `tf` conversion in new v6 models

### LINKED ISSUE #5147: Can't build TF saved_model or tflite with new v6
## 🐛 Bug

Can't build TF saved_model or tflite with new v6
## To Reproduce (REQUIRED)
Run the google colab- only need to run the below block:
```
# CI Checks
%%shell
export PYTHONPATH="$PWD"  # to run *.py. files in subdirectories
rm -rf runs  # remove runs/
for m in yolov5s; do  # models
  python train.py --weights $m.pt --epochs 3 --img 320 --device 0  # train pretrained
  python train.py --weights '' --cfg $m.yaml --epochs 3 --img 320 --device 0  # train scratch
  for d in 0; do # cpu; do  # devices
    python detect.py --weights $m.pt --device $d  # detect official
    python detect.py --weights runs/train/exp/weights/best.pt --device $d  # detect custom
    python val.py --weights $m.pt --device $d # val official
    python val.py --weights runs/train/exp/weights/best.pt --device $d # val custom
  done
  python hubconf.py  # hub
  python models/yolo.py --cfg $m.yaml  # build PyTorch model
  python models/tf.py --weights $m.pt  # build TensorFlow model
  python export.py --img 128 --batch 1 --weights $m.pt --include tflite #torchscript onnx  # export
done
```
output:
missing SPPF in tf.py

## Environment

google colab + gpu

## Some debug:
For start the SPPF is missing from `tf.py`:

```
from models.common import Conv, Bottleneck, SPP, DWConv, Focus, BottleneckCSP, Concat, autopad, C3, **SPPF**

...

class TFSPPF(keras.layers.Layer):
    # Spatial pyramid pooling-Fast layer
    def __init__(self, c1, c2, k=5, w=None):
        super(TFSPPF, self).__init__()
        c_ = c1 // 2  # hidden channels
        self.cv1 = TFConv(c1, c_, 1, 1, w=w.cv1)
        self.cv2 = TFConv(c_ * 4, c2, 1, 1, w=w.cv2)
        self.m = keras.layers.MaxPool2D(pool_size=k, strides=1, padding='SAME')

    def call(self, inputs):
        x = self.cv1(inputs)
        y1 = self.m(x)
        y2 = self.m(y1)
        return self.cv2(tf.concat([x, y1, y2, self.m(y2)], 3))
```

After building the TFSPPF class we are still getting this new error:
```
Eager execution of tf.constant with unsupported shape (value has 131072 elements, shape is (1, 1, 512, 512) with 262144 elements)
```

### PR DESCRIPTION
**Pull request by YoniChechik**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Integration of TFSPPF, a new TensorFlow spatial pyramid pooling-fast layer, into the YOLOv5 architecture.

### 📊 Key Changes
- 🔧 Added the `TFSPPF` class to the TensorFlow model file, providing a new layer for spatial pyramid pooling with a fast implementation.
- 🛠️ Updated the `parse_model` function to include the new `SPPF` layer when constructing models.
- 💼 Rearranged the import statements in `models.tf.py` to include the newly added `SPPF` alongside other common model components.

### 🎯 Purpose & Impact
- 🚀 The new `TFSPPF` layer is designed to improve model performance by providing a more efficient way to handle spatial data.
- 🆕 By integrating this layer into YOLOv5's TensorFlow implementation, developers and users can take advantage of enhanced object detection capabilities.
- 🔍 The potential impact includes faster inference times and possibly improved accuracy, benefiting a wide range of applications from autonomous driving to security surveillance.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@YoniChechik PR is merged. Thank you for your contributions to YOLOv5 🚀 and Vision AI ⭐

**Comment by rahuja23** [comment]
When I am exporting a custom trained model into tflite and then loading it , I am getting the following error:
 Model provided has model identifier '
Any helps or suggestions what can be done?

**Comment by glenn-jocher** [comment]
@rahuja23 👋 hi, thanks for letting us know about this possible problem with YOLOv5 🚀. We've created a few short guidelines below to help users provide what we need in order to get started investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

* ✅ **Minimal** – Use as little code as possible that still produces the same problem
* ✅ **Complete** – Provide **all** parts someone else needs to reproduce your problem in the question itself
* ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

In addition to the above requirements, for [Ultralytics](https://ultralytics.com/) to provide assistance your code should be:

* ✅ **Current** – Verify that your code is up-to-date with current GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been resolved by previous commits.
* ✅ **Unmodified** – Your problem must be reproducible without any modifications to the codebase in this repository. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all of the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) and providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by rahuja23** [comment]
torch  - 1.9.1 
torchvision -  0.10.1
tensorflow- 2.6.0
I cloned the recent repo and trained the model on  cloud using single GPU instance  and when trying to convert the model with the latest export.py I am getting the following exception:
`[34m[1mTensorFlow saved_model:[0m export failure: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.`

**Comment by glenn-jocher** [comment]
@rahuja23 your error is not reproducible. Training and export to TFLite operates correctly.
```python
# Train YOLOv5s on COCO128 for 3 epochs
!python train.py --img 640 --batch 16 --epochs 3 --data coco128.yaml --weights yolov5s.pt --cache

# Export to TFLite, SUCCESS
!python export.py --weights runs/train/exp/weights/best.pt --include tflite
```

Screen Shot 2021-10-15 at 12 09 23 PM

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

* ✅ **Minimal** – Use as little code as possible that still produces the same problem
* ✅ **Complete** – Provide **all** parts someone else needs to reproduce your problem in the question itself
* ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

In addition to the above requirements, for [Ultralytics](https://ultralytics.com/) to provide assistance your code should be:

* ✅ **Current** – Verify that your code is up-to-date with current GitHub [master](https://github.com/ultralytics/yolov5/tree/master), and if necessary `git pull` or `git clone` a new copy to ensure your problem has not already been resolved by previous commits.
* ✅ **Unmodified** – Your problem must be reproducible without any modifications to the codebase in this repository. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all of the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/yolov5/issues/new/choose) and providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by rahuja23** [comment]
Ok I figured it out I was setting the device parameter in export file as well to CUDA,  if we leave it to default it works. Thanks for the new push guys 👍🏻
